### PR TITLE
Add create_conv_retrieval_chain func

### PR DESCRIPTION
**Pull request by nfcampos**
```
                                                     +----------+
                                                     | MapInput |
                                                   **+----------+****
                                               ****                  ****
                                           ****                          ***
                                         **                                 ****
                  +------------------------------------+                        **
                  | Lambda(itemgetter('chat_history')) |                         *
                  +------------------------------------+                         *
                                     *                                           *
                                     *                                           *
                                     *                                           *
                       +---------------------------+            +--------------------------------+
                       | Lambda(_get_chat_history) |            | Lambda(itemgetter('question')) |
                       +---------------------------+            +--------------------------------+
                                     *                                           *
                                     *                                           *
                                     *                                           *
                      +----------------------------+                +------------------------+
                      | ContextSet('chat_history') |                | ContextSet('question') |
                      +----------------------------+                +------------------------+
                                               ****                  ****
                                                   ****          ****
                                                       **      **
                                                     +-----------+
                                                     | MapOutput |
                                                     +-----------+
                                                           *
                                                           *
                                                           *
                                                  +----------------+
                                                  | PromptTemplate |
                                                  +----------------+
                                                           *
                                                           *
                                                           *
                                                    +-------------+
                                                    | FakeListLLM |
                                                    +-------------+
                                                           *
                                                           *
                                                           *
                                                  +-----------------+
                                                  | StrOutputParser |
                                                  +-----------------+
                                                           *
                                                           *
                                                           *
                                            +----------------------------+
                                            | ContextSet('new_question') |
                                            +----------------------------+
                                                           *
                                                           *
                                                           *
                                                +---------------------+
                                                | SequentialRetriever |
                                                +---------------------+
                                                           *
                                                           *
                                                           *
                                        +------------------------------------+
                                        | Lambda(_reduce_tokens_below_limit) |
                                        +------------------------------------+
                                                           *
                                                           *
                                                           *
                                           +-------------------------------+
                                           | ContextSet('input_documents') |
                                           +-------------------------------+
                                                           *
                                                           *
                                                           *
                                                     +----------+
                                                  ***| MapInput |****
                                           *******   +----------+    ********
                                   ********                *                 *******
                            *******                         *                       ********
                        ****                                *                               ****
+-------------------------------+            +----------------------------+            +----------------------------+
| ContextGet('input_documents') |            | ContextGet('chat_history') |            | ContextGet('new_question') |
+-------------------------------+****        +----------------------------+            +----------------------------+
                                     *********                *                 *******
                                              ********         *          ******
                                                      *****    *      ****
                                                         +-----------+
                                                         | MapOutput |
                                                         +-----------+
                                                                *
                                                                *
                                                                *
                                                        +-------------+
                                                        | FakeListLLM |
                                                        +-------------+
                                                                *
                                                                *
                                                                *
                                                          +----------+
                                                       ***| MapInput |***
                                               ********   +----------+   ******
                                        *******                 *              *****
                                ********                        *                   ******
                            ****                                *                         ***
    +-------------------------------+            +----------------------------+            +-------------+
    | ContextGet('input_documents') |            | ContextGet('new_question') |          **| Passthrough |
    +-------------------------------+            +----------------------------+   *******  +-------------+
                                     *******                 *              ******
                                            ******           *       *******
                                                  ****      *    ****
                                                     +-----------+
                                                     | MapOutput |
                                                     +-----------+
```

### CODE REVIEW

File: libs/core/langchain_core/language_models/base.py [hunk 1]
```diff
@@ -49,11 +49,13 @@ def _get_token_ids_default_method(text: str) -> List[int]:
...
 LanguageModelInput = Union[PromptValue, str, List[BaseMessage]]
-LanguageModelOutput = TypeVar("LanguageModelOutput")
+LanguageModelOutput = Union[BaseMessage, str]
+LanguageModelLike = Runnable[LanguageModelInput, LanguageModelOutput]
+LanguageModelOutputVar = TypeVar("LanguageModelOutputVar", BaseMessage, str)
 
 
 class BaseLanguageModel(
-    RunnableSerializable[LanguageModelInput, LanguageModelOutput], ABC
+    RunnableSerializable[LanguageModelInput, LanguageModelOutputVar], ABC
```
Comment (jacoblee93) [inline libs/core/langchain_core/language_models/base.py:58]: Should this be in a separate PR?

File: libs/langchain/langchain/chains/conversational_retrieval/base.py [hunk 2]
```diff
@@ -465,3 +470,68 @@ def from_llm(
...
+    combine_docs_chain: Runnable[Dict[str, Any], Dict[str, Any]],
+    max_tokens_limit: Optional[int] = None,
+) -> Runnable:
+    if max_tokens_limit is not None:
+        assert max_tokens_limit > 0
+    if combine_docs_chain.input_schema.__custom_root_type__:
+        raise ValueError("`combine_docs_chain` must accept dict as input.")
+    if combine_docs_chain.output_schema.__custom_root_type__:
+        raise ValueError("`combine_docs_chain` must return dict as output.")
+    if set(question_gen_prompt.input_variables) != {"question", "chat_history"}:
```
Comment (jacoblee93) [inline libs/langchain/langchain/chains/conversational_retrieval/base.py:488]: Do we also need to check `input_documents` here as well?

File: libs/langchain/langchain/chains/conversational_retrieval/base.py [hunk 3]
```diff
@@ -465,3 +470,68 @@ def from_llm(
...
+            tokens = [get_num_tokens(doc.page_content) for doc in docs]
+            token_count = sum(tokens[:num_docs])
+            while token_count > max_tokens_limit:
+                num_docs -= 1
+                token_count -= tokens[num_docs]
+
+        return docs[:num_docs]
+
+    return (
+        RunnableMap(
```
Comment (jacoblee93) [inline libs/langchain/langchain/chains/conversational_retrieval/base.py:516]: Should name these runs nicely

**Comment by Oudinunit** [comment]
Ok I come back home

File: libs/langchain/langchain/chains/conversational_retrieval/base.py [hunk 4]
```diff
@@ -465,3 +467,59 @@ def from_llm(
             callbacks=callbacks,
             **kwargs,
         )
+
+
+def create_chat_retriever_chain(
```
Comment (jacoblee93) [inline libs/langchain/langchain/chains/conversational_retrieval/base.py:472]: Why `chat` instead of `conversational`?

Comment (jacoblee93) [reply libs/langchain/langchain/chains/conversational_retrieval/base.py:472]: My (perhaps dated) mental model was `chat == ChatModels`, `conversational == chat history`

File: libs/langchain/tests/unit_tests/retrievers/parrto_retriever.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from typing import List
```
Comment (jacoblee93) [inline libs/langchain/tests/unit_tests/retrievers/parrto_retriever.py:1]: Typo in filename

File: libs/langchain/langchain/chains/conversational_retrieval/base.py [hunk 6]
```diff
@@ -465,3 +467,59 @@ def from_llm(
...
+    llm: LanguageModelLike,
+    retriever: RetrieverLike,
+    prompt: BasePromptTemplate,
+) -> RetrieverOutputLike:
```
Comment (jacoblee93) [inline libs/langchain/langchain/chains/conversational_retrieval/base.py:476]: Type seems too loose? Any reason it can't be:

```python
Runnable[Dict[str, Any], RetrieverOutput]
```

File: libs/langchain/langchain/chains/conversational_retrieval/base.py [hunk 7]
```diff
@@ -465,3 +467,59 @@ def from_llm(
...
+        raise ValueError(
+            "Expected `input` to be a prompt variable, "
+            f"but got {prompt.input_variables}"
+        )
+    retrieve_documents = RunnableBranch(
+        (
+            lambda x: len(x.get("chat_history", [])) == 0,
+            (lambda x: x["input"]) | retriever,
+        ),
+        prompt | llm | StrOutputParser() | retriever,
```
Comment (jacoblee93) [inline libs/langchain/langchain/chains/conversational_retrieval/base.py:523]: So `chat_history` would have to be strictly a string or a list of messages depending on the prompt right?

File: libs/langchain/langchain/chains/conversational_retrieval/base.py (hunk 4 above)
Comment (hwchase17) [reply libs/langchain/langchain/chains/conversational_retrieval/base.py:472]: will change to `create_chat_history_retriever`

File: libs/langchain/langchain/chains/conversational_retrieval/base.py (hunk 7 above)
Comment (hwchase17) [reply libs/langchain/langchain/chains/conversational_retrieval/base.py:523]: yeah

people can always add to start if they way?

Comment (jacoblee93) [reply libs/langchain/langchain/chains/conversational_retrieval/base.py:526]: Yeah not opposed, just want to make sure
